NOISSUE - Add pure JSON transformer

manuio opened the pull request:
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* JSON key-value pairs of the message are transformed as a single Mainflux message where "key" and "value" will be represented as "name" and "Value, StringValue, DataValue or BoolValue". key-value pairs can be also filtered. 

* For example, this JSON message:
```
{"device":"1234", "type": "temperature", "data": {"value": 1, "ts": 12345678}}
``` 

will be converted in 4 transformers.Message:
```
{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "type",
  "time": 1602499859.975915,
  "string_value": "temperature"
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "device",
  "time": 1602499859.975915,
  "string_value": "1234"
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "data/value",
  "time": 1602499859.975915,
  "value": 1
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "data/ts",
  "time": 1602499859.975915,
  "value": 12345678
}
```

mteodor commented on docker/addons/cassandra-writer/filters.toml:
```diff
@@ -0,0 +1,11 @@
+filter = ["channels.>"]
+
+# If writer is used with ContentType application/json JSON key-value pairs can
+# be filtered. Keys are used to store correspondinng pair. (e.g ["temperature", "humidity"]).
```
this means that only the enlisted keys will be saved, right?

manuio replied:
Right! Or all if listed key is "*"

mteodor commented on pkg/transformers/json/transformer.go:
```diff
@@ -0,0 +1,92 @@
+
+	for k, v := range payload {
+		// Apply key filter
+		for _, tk := range t.keys {
```
how about using a map for searching?

mteodor commented on pkg/transformers/json/transformer.go:
```diff
@@ -0,0 +1,92 @@
+		Subtopic:  mm.Subtopic,
+		Publisher: mm.Publisher,
+		Protocol:  mm.Protocol,
+		Name:      key,
```
is there a way to group records by incoming message, as I understand incoming JSON payload that has like 10 fields will produce 10 records in writers database, but do we have a way to group these 10 records?

manuio replied:
Writers use`transformers.Message` format, the best that you can do is store the JSON as `DataValue` or `StringValue`

mteodor replied:
what I meant is that you can easily determine that group of 10 records is from one `JSON` message.

manuio replied:
Yes, for example with a reader JSON transformer that will group messages by timestamp (kind of state) and re-create the corresponding JSON

drasko commented on docker/addons/cassandra-writer/filters.toml:
```diff
@@ -0,0 +1,12 @@
+# To listen all messsage broker subjects use default value "channels.>".
```
If we start adding stuff, I would prefer calling this `config.toml`, and then separate tables for `nats` and `payload`

manuio replied:
I agree

drasko commented on pkg/transformers/message.go:
```diff
@@ -1,4 +1,4 @@
-package senml
+package transformers
```
I am not sure that `Message{}` should be in `transformers` package.

Maybe it should not be in `senml` neither, but I am really skeptical about this one.

manuio replied:
Let's think about. This is more a DB message format.

manuio committed:
NOISSUE - Add pure JSON transformer

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

dborovcanin committed:
MF-1254 - Create universal JSON writer (#1260)

* Update JSON transformer

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Use interface{} instead of Transformer message

Using interface{} says nothing, but it's generic and handling specific
transformed message is offloaded to the Repository. This leaves an
option for more generic Repositories and simpler integration. Also,
Repository is no more dependent on SenML Transformer, but its specific
implementation is.

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update InfluxDB writer to support JSON

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix collection/measurement/table naming

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix default transformer for InfluxDB writer

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Add support for JSON in Postgres Writer

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Add JSON support to MongoDB Writer

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Add JSON transformer tests

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Replace JSON nested objects separator

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix JSON transformer test

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Rename configuration files and settings

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Remove keys filter

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix configPath naming in InfluxDB writer

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix Cassandra Writer README

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix PR remarks

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
